http-webhook.md

```diff
@@ -0,0 +1,339 @@
+field or the HTTP request entity-body. All further caveats cited in
+[RFC6750][RFC6750] apply equivalently.
+
+## 4. Abuse Protection
+
+Any system that allows registration of and delivery of notifications to
+arbitary HTTP endpoints can potentially be abused such that someone maliciously
+or inadvertently registers the address of a system that does not expect such
+requests and for which the registering party is not authorized to perform such
+a registration. In extreme cases, a notification infrastructure could be abused
+to launch denial-of-service attacks against an arbitrary web-site.
+
+To protect the sender from being abused in such a way, a legitimate delivery
+target needs to indicate that it agrees with notifications being delivered to it.
+
+Reaching the delivery agreement is realized using the following validation
+handshake. The handshake can either be executed immediately at registration
+time or as a "pre-flight" request immediately preceding a delivery. If the
+handshake outcome is that the delivery is denied, the sender MUST NOT deliver
+events to the target.
+
+It is important to understand is that the handshake does not aim to establish
+an authentication or authorization context. It only serves to protect the
+sender from being told to a push to a destination that is not expecting the
+traffic. While this specification mandates use of an authorization model, this
+mandate is not sufficient to protect any arbitrary website from unwanted
+traffic if that website doesn't implement access control and therefore ignores
+the `Authorization` header.
+
+Delivery targets SHOULD support the abuse protection feature. If a target does
+not support the feature, the sender MAY choose not to send to the target, at
+all, or send only at a very low request rate.
+
+### 4.1.1. Validation request
+
+The validation request uses the HTTP [OPTIONS][OPTIONS] method. The request
+is directed to the exact resource target URI that is being registered.
+
+With the validation request, the sender asks the target for permission to
+send notifications, and it can declare a desired request rate (requests per
+minute).
+
+The delivery target will respond with a permission statement and the permitted
+request rate.
+
+The following header fields MUST be included in the validation request:
+
+#### 4.1.1.2. WebHook-Request-Origin
+
+The `WebHook-Request-Origin` header requests permission to send notifications
+from this sender, and contains a DNS expression that identifies the sending
+system, for example "eventemitter.example.com". The value is meant to summarily
+identify all sender instances that act on the behalf of a certain system, and
+not an individual host.
+
+After the handshake and when permission is granted, the sender MUST use the
+`Origin` request header for each delivery request, with the value matching that
+of this header.
+
+For example:
+
+``` text
+WebHook-Request-Origin: eventemitter.example.com
+```
+
+#### 4.1.1.2. WebHook-Request-Rate
```

Comment:
Rate limiting might apply to the endpoint (e.g. the Tweet analyzer subsystem can only support 1500QPS) or to a client as a whole (each twitter user can only send 50req/hr). I wonder if we need an additional header to indicate the scope on which the rate limit applies.

Reply:
The request rate applies to this exact target resource. If the resource has some composite rules, it ought to do the math and give the sender a simple answer rather than complicating matters by giving two answers that may be conflicting. Keep it simple.